INSTITUTO TECNOLÓGICO Y DE ESTUDIOS SUPERIORES DE MONTERREY

MNA - Maestría Inteligencia Artificial Aplicada

## Tarea 3: Aplicación de Algoritmos de Aprendizaje Supervisado y No Supervisado con PySpark
#### **A01795907 - Samantha Rizo Franco**

**Asignatura:** Análisis de Grandes Volúmenes de Datos  
**Dataset:** Airbnb Global Listings (Kaggle)  
**Herramienta principal:** Apache Spark — PySpark MLlib  


# 1. Introducción teórica

## 1.1 Big Data y procesamiento distribuido

El término **Big Data** hace referencia a conjuntos de datos cuya magnitud, velocidad de generación y variedad exceden las capacidades de procesamiento de los sistemas tradicionales. Se caracteriza por *5 Vs*: volumen, velocidad, variedad, veracidad y valor. Para procesar eficientemente estos datos se utilizan arquitecturas distribuidas, donde tanto el almacenamiento como el procesamiento se reparten entre múltiples nodos de un clúster.

**Apache Spark** es uno de los frameworks más utilizados para el procesamiento distribuido debido a su capacidad de ejecutar operaciones en memoria, reduciendo significativamente los tiempos de procesamiento frente a enfoques tradicionales como MapReduce. Spark proporciona estructuras de datos distribuidas como los *Resilient Distributed Datasets* (RDDs) y los *DataFrames*, además de bibliotecas especializadas para análisis de datos y aprendizaje automático.

## 1.2 Machine Learning

El **Machine Learning (ML)** llamado en español **aprendizaje automático** es una rama de la inteligencia artificial que desarrolla algoritmos capaces de aprender patrones a partir de los datos para realizar predicciones o descubrir estructuras ocultas sin necesidad de programar explícitamente todas las reglas de decisión.

Integrada de manera nativa en Apache Spark tenemos la biblioteca **MLlib**, diseñada para entrenar modelos de aprendizaje automático de manera distribuida sobre grandes volúmenes de información, permitiendo escalar tanto los procesos de entrenamiento como de inferencia [2].

## 1.3 Preprocesamiento y análisis descriptivo de datos

Previo a entrenar un modelo de aprendizaje automático es necesario realizar procesos de **limpieza, transformación y exploración de datos** para garantizar la calidad de los datos y mejorar el desempeño de los modelos de aprendizaje automático. Como por ejemplo manejo de valores faltantes, codificación de variables categóricas, normalización o escalamiento de variables numéricas, generación de nuevas características (*feature engineering*) y análisis estadístico descriptivo de los datos [1].

PySpark proporciona herramientas para realizar estas tareas como:

- **VectorAssembler**: combina múltiples variables en un único vector de características.
- **StringIndexer**: transforma variables categóricas en índices numéricos.
- **OneHotEncoder**: representa variables categóricas mediante vectores binarios.
- **VectorIndexer**: identifica automáticamente variables categóricas dentro de un vector de características.
- **StandardScaler** y **MinMaxScaler**: normaliza variables numéricas.
- **MaxAbsScaler** y **RobustScaler**: escala variables utilizando diferentes estrategias de normalización.
- **Imputer**: reemplaza valores faltantes mediante estadísticas como la media o la mediana.
- **Binarizer**: convierte variables numéricas en variables binarias utilizando un umbral.
- **Bucketizer** y **QuantileDiscretizer**: discretiza variables continuas en intervalos o categorías.
- **PCA (Principal Component Analysis)**: reduce la dimensionalidad de los datos conservando la mayor cantidad posible de información.
- **Tokenizer**, **RegexTokenizer**, **StopWordsRemover**, **CountVectorizer** y **HashingTF**: para el procesamiento y transformación de datos textuales.
- Así como funciones estadísticas para calcular correlaciones, distribuciones, frecuencias y estadísticas descriptivas de los datos [1,4].

## 1.4 Aprendizaje supervisado

El **aprendizaje supervisado** utiliza un conjunto de datos etiquetado, donde cada observación posee una variable objetivo conocida. El objetivo es aprender a predecir dicha variable para nuevas observaciones [2].

Los principales modelos de aprendizaje supervisado son:

- **Clasificación:** predicción de categorías o clases discretas.
- **Regresión:** predicción de valores continuos.

A continuación los algoritmos supervisados disponibles en PySpark MLlib [5]:

#### Clasificación

- Regresión logística: **LogisticRegression**
- Árboles de Decisión: **DecisionTreeClassifier**
- Random Forest: **RandomForestClassifier**
- Gradient Boosted Trees: **GBTClassifier**
- Máquinas de Soporte Vectorial (SVM): **LinearSVC**
- Naive Bayes: **NaiveBayes**
- Redes Neuronales Artificiales: **MultilayerPerceptronClassifier**

También tenemos la herramienta **OneVsRest** que nos sirve para convertir un problema de clasificación multiclase en varios problemas de clasificación binaria.

#### Regresión

- Regresión Lineal: **LinearRegression**
- Regresión Lineal Generalizada: **GeneralizedLinearRegression**
- Árboles de Decisión: **DecisionTreeRegressor**
- Random Forest: **RandomForestRegressor**
- Gradient Boosted Trees: **GBTRegressor**
- Regresión Isotónica: **IsotonicRegression**

## 1.5 Aprendizaje no supervisado

Por otro lado, el **aprendizaje no supervisado** trabaja con datos que no poseen etiquetas conocidas. Su objetivo es descubrir patrones, agrupamientos o estructuras ocultas presentes en los datos [2].

Las principales aplicaciones del aprendizaje no supervisado incluyen:

- Segmentación o agrupamiento (*clustering*).
- Detección de anomalías.
- Reducción de dimensionalidad.
- Descubrimiento de tópicos en colecciones de texto.

A continuación, algoritmos no supervisados disponibles en PySpark MLlib [6]:

#### Clustering
- K-Means: **KMeans**,
- Clustering Bisecting K-Means: **BisectingKMeans**,
- Modelos de mezcla Gaussiana: **GaussianMixture**,
- Clustering basado en grafos: **PowerIterationClustering (PIC)**,
- Clustering probabilístico: **LatentDirichletAllocation (LDA)**.

#### Reducción de dimensionalidad
- Análisis de Componentes Principales: **PCA**.

#### Modelos probabilísticos y tópicos
- Latent Dirichlet Allocation: **LDA**

#### Detección de anomalías
PySpark MLlib no tiene un algoritmo específico dedicado exclusivamente a anomalías, pero se construye combinando otros modelos.

## 1.6 Importancia de PySpark para el análisis a gran escala

PySpark constituye la interfaz de programación en Python para Apache Spark que combina la simplicidad de Python con la capacidad de procesamiento distribuido de Spark. Sus principales características son:

- Procesamiento distribuido y tolerante a fallos.
- Ejecución en memoria para acelerar cálculos.
- Integración con SQL, procesamiento de flujos (*streaming*) y aprendizaje automático.
- Escalabilidad horizontal sobre clústeres de cómputo.
- Integración con sistemas de almacenamiento distribuidos.

Convirtiendolo en una herramienta idónea para proyectos de ciencia de datos y aprendizaje automático sobre conjuntos de datos de gran tamaño.



In [1]:
# INSTALAR JAVA (OBLIGATORIO PARA SPARK)
!apt-get update -qq
!apt-get install openjdk-11-jdk-headless -qq > /dev/null

import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-11-openjdk-amd64"

# INSTALAR PYSPARK
!pip install -q pyspark==3.5.1

# IMPORTAR LIBRERÍAS
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark.sql.functions import col, when, count, concat_ws, log1p
from pyspark.sql.types import DoubleType

from pyspark.ml import Pipeline
from pyspark.ml.feature import (
    StringIndexer, StringIndexerModel,
    OneHotEncoder, VectorAssembler,
    StandardScaler, IndexToString
)
from pyspark.ml.classification import (
    RandomForestClassifier,
    GBTClassifier,
    OneVsRest
)
from pyspark.ml.clustering import KMeans
from pyspark.ml.evaluation import (
    MulticlassClassificationEvaluator,
    ClusteringEvaluator
)

# CREAR SPARK SESSION
spark = (
    SparkSession.builder
    .master("local[*]")
    .appName("Tarea3_ML_Airbnb")
    .config("spark.sql.repl.eagerEval.enabled", True)
    .config("spark.driver.memory", "4g")
    .config("spark.sql.shuffle.partitions", "8")
    .getOrCreate()
)
spark

# MONTAR DRIVE
from google.colab import drive
drive.mount('/content/drive')

# 2. Selección de los datos

### 2.1 Origen del dataset

Para esta investigación se utilizó el dataset “Airbnb Global Listings” [7], obtenido de Kaggle, el cual contiene aproximadamente 494,954 registros y 89 variables relacionadas con propiedades publicadas en Airbnb alrededor del mundo; incluyendo atributos descriptivos de la propiedad (tipo de habitación, tipo de inmueble, número de habitaciones, baños y camas), atributos económicos (precio diario, semanal, mensual, depósito de seguridad y tarifa de limpieza), variables geográficas (país, ciudad, vecindario, latitud y longitud), métricas de reseñas (puntuaciones, número de reseñas) y características del anfitrión. Se distribuye como un archivo tabular en formato CSV de aproximadamente **1.8 GB**, un tamaño que ya se considera adecuado para trabajar con herramientas de Big Data como PySpark.

### 2.2 Problemas detectados durante la carga

Durante la lectura del archivo se identificaron varios problemas:

1. **Saltos de línea internos** en campos de texto que rompían la estructura tabular, lo cual se resolvió con `multiLine=True`.
2. **Variables numéricas almacenadas como cadenas de texto** (por ejemplo, `Price`, `Bathrooms`, `Bedrooms`), por lo cual se aplicaron procesos de conversión de tipos de datos.
3. **Alta proporción de valores nulos** en variables como `Square Feet`, `Cleaning Fee` y `Review Scores Rating`.
4. **Alta cardinalidad** en variables geográficas (`city`, `neighbourhood`).
5. **Categorías inválidas** en variables como `Room Type` (valores como `"9"` o nulos).

In [2]:
# Carga del dataset
ruta = '/content/drive/MyDrive/Analisis_de_grandes_volumenes_de_datos/airbnb-listings.csv'

df = spark.read.csv(
    ruta,
    header=True,
    sep=';',
    multiLine=True,
    quote='"',
    escape='"',
    inferSchema=True
)

print("Registros originales:", df.count())
print("Columnas originales: ", len(df.columns))

Registros originales: 494954
Columnas originales:  89


### 2.3 Justificación de la selección de variables

Para los modelos se seleccionaron las variables con mayor poder informativo respecto al precio y a las características estructurales de las propiedades: `Room Type`, `Property Type`, `Bedrooms`, `Bathrooms`, `Accommodates`, `Review Scores Rating`, `Availability 365`, `Cleaning Fee` (r=0.55), `Beds` (r=0.29), `Extra People` y `Price`. Esta elección se respalda con el análisis exploratorio previo, en el cual se detectaron correlaciones positivas entre el precio y los atributos de capacidad, además de la relevancia del tipo de habitación como factor diferenciador del segmento de mercado.

### 2.4 Método de muestreo: estratificado

Debido a la magnitud del dataset y al desbalance de algunas categorías, se utilizó un **muestreo estratificado**. Este método consiste en dividir la población en estratos homogéneos respecto a una o varias variables de interés y extraer una fracción de cada estrato, garantizando así la representatividad de todas las subpoblaciones en la muestra final.

En el proyecto previo se construyeron **9 particiones** $P_1, \ldots, P_9$ mediante el cruce de las tres categorías de `Room Type` (`Entire home/apt`, `Private room`, `Shared room`) y las tres categorías de `price_range` (`Low`, `Mid`, `High`). Sobre cada partición se aplicó un muestreo aleatorio simple con fracción del 5% y semilla fija (`seed=42`), obteniendo así submuestras $M_1, \ldots, M_9$, cuya unión conforma la **muestra estratificada M** utilizada. De tal forma que el muestreo estratíficado nos permite:
- Preservar la proporción de las categorías minoritarias (`Shared room` y `price_range = High`), cuya pérdida en un muestreo aleatorio simple sesgaría los modelos hacia las clases mayoritarias.
- Reducir la varianza del estimador al asegurar cobertura completa de los estratos relevantes (estimaciones de desempeño más estables).
- Facilitar el entrenamiento de modelos de clasificación multiclase al mantener la variable objetivo `price_range`, evitando el problema de clases prácticamente ausentes en el conjunto de entrenamiento.

## 3. Preparación de los datos

La preparación de los datos es un paso fundamental en el Machine Learning debido a que la calidad de los datos con que se alimenta un modelo, limita la calidad del modelo. A continuación los pasos de preprocesamiento aplicados.

### 3.1 Eliminación de registros inválidos

Durante la exploración se detectaron valores espurios en `Room Type` (por ejemplo, 9 y valores nulos) que no corresponden a categorías válidas de la plataforma Airbnb. Estos registros fueron eliminados conservando únicamente las categorías válidas, en este caso: `Entire home/apt`, `Private room` y `Shared room`. También se filtraron registros con `Property Type` o `Price` nulo, porque es indispensable tanto para construir la variable objetivo.

### 3.2 Tratamiento de valores nulos

Los valores nulos en variables predictoras numéricas (`Bedrooms`, `Bathrooms`, `Accommodates`, `Review Scores Rating`, `Availability 365`) se imputaron utilizando su respectiva **mediana**. Se eligió la mediana en lugar de la media porque las distribuciones presentan sesgo a la derecha y valores atípicos, lo cual distorsionaría la media. Para variables categóricas (`Property Type`), se reemplazaron por la categoría `"Unknown"`.

### 3.3 Conversión de tipos de datos

Se utilizó `inferSchema=True` resultando en una inferencia inicial razonable, peor fue necesario forzar el tipado de las variables numéricas a `DoubleType` para garantizar la compatibilidad con los transformadores y estimadores de MLlib, los cuales requieren entradas numéricas en precisión doble para el ensamblado del vector de características.

### 3.4 Construcción de la variable categórica price_range

A partir de los cuantiles empíricos del precio ($p_{25} = 55$ USD y $p_{75} = 150$ USD), se discretizó la variable continua `Price` en tres categorías:

- **Low:** $\text{Price} < p_{25}$  
- **Mid:** $p_{25} \leq \text{Price} \leq p_{75}$  
- **High:** $\text{Price} > p_{75}$

Transformando un problema de regresión en uno de clasificación multiclase.

### 3.5 Detección de inconsistencias

Se verificó que `Price`, `Bedrooms`, `Bathrooms` y `Accommodates` no tuvieran valores negativos, descartando registros con valores anómalos. Pro otro lado, se inspeccionaron los dominios de variables categóricas con `distinct()` para detectar codificaciones erróneas.



In [3]:
# Limpieza de datos
df_clean = (
    df.filter(col("Room Type").isin("Entire home/apt", "Private room", "Shared room"))
      .filter(col("Property Type").isNotNull())
      .filter(col("Price").isNotNull())
)

# Variables numéricas
#   Cleaning Fee (r=0.55), Beds (r=0.29), Extra People (r=0.27),
#   Guests Included (r=0.19)
num_cols = ["Price", "Bedrooms", "Bathrooms", "Accommodates", "Beds", "Cleaning Fee", "Extra People", "Guests Included",
    "Number of Reviews", "Host Listings Count","Review Scores Rating", "Availability 365"]
for c in num_cols:
    df_clean = df_clean.withColumn(c, col(c).cast(DoubleType()))

# Filtrar valores imposibles
df_clean = (
    df_clean
    .filter(col("Price") > 0)
    .filter(col("Accommodates") > 0)
    .filter(col("Bedrooms") >= 0)
    .filter(col("Bathrooms") >= 0)
)

# Imputación de nulos numéricos con la mediana
medianas = {c: df_clean.approxQuantile(c, [0.5], 0.01)[0] for c in num_cols}
print("Medianas calculadas:", medianas)
for c, m in medianas.items():
    df_clean = df_clean.fillna({c: m})

# Imputación categórica
df_clean = df_clean.fillna({
    "Property Type": "Unknown",
    "Cancellation Policy": "Unknown",
    "Bed Type": "Unknown"
})

# Feature engineering: log1p(Number of Reviews) para reducir sesgo de larga cola
df_clean = df_clean.withColumn("log_reviews", log1p(col("Number of Reviews")))

print("Registros tras limpieza:", df_clean.count())

Medianas calculadas: {'Price': 90.0, 'Bedrooms': 1.0, 'Bathrooms': 1.0, 'Accommodates': 2.0, 'Beds': 1.0, 'Cleaning Fee': 40.0, 'Extra People': 0.0, 'Guests Included': 1.0, 'Number of Reviews': 4.0, 'Host Listings Count': 1.0, 'Review Scores Rating': 95.0, 'Availability 365': 142.0}
Registros tras limpieza: 484813


In [4]:
# Construcción de price_range a partir de cuantiles

p25, p75 = df_clean.approxQuantile("Price", [0.25, 0.75], 0.01)
print(f"p25 = {p25} | p75 = {p75}")

df_clean = df_clean.withColumn(
    "price_range",
    when(col("Price") < p25, "Low")
    .when((col("Price") >= p25) & (col("Price") <= p75), "Mid")
    .otherwise("High")
)

df_clean.groupBy("price_range").count().orderBy("price_range").show()

df_clean = df_clean.withColumn(
    "stratum",
    concat_ws("_",
        col("Room Type"),
        col("price_range")
    )
)

df_clean.groupBy("stratum").count().orderBy("stratum").show(truncate=False)

p25 = 55.0 | p75 = 150.0
+-----------+------+
|price_range| count|
+-----------+------+
|       High|115780|
|        Low|115182|
|        Mid|253851|
+-----------+------+

+--------------------+------+
|stratum             |count |
+--------------------+------+
|Entire home/apt_High|102747|
|Entire home/apt_Low |27490 |
|Entire home/apt_Mid |174834|
|Private room_High   |12446 |
|Private room_Low    |81423 |
|Private room_Mid    |77002 |
|Shared room_High    |587   |
|Shared room_Low     |6269  |
|Shared room_Mid     |2015  |
+--------------------+------+



In [5]:
estratos = (
    df_clean
    .select("stratum")
    .distinct()
    .rdd
    .flatMap(lambda x: x)
    .collect()
)

# Fracción aumentada a 10% para mayor representatividad y mejor accuracy
fractions = {e: 0.10 for e in estratos}

muestra_M = df_clean.sampleBy(
    "stratum",
    fractions=fractions,
    seed=42
)

muestra_M = muestra_M.cache()
print("Tamaño muestra M:", muestra_M.count())


Tamaño muestra M: 48369


In [6]:
muestra_M.groupBy("stratum") \
         .count() \
         .orderBy("stratum") \
         .show(truncate=False)

+--------------------+-----+
|stratum             |count|
+--------------------+-----+
|Entire home/apt_High|10169|
|Entire home/apt_Low |2751 |
|Entire home/apt_Mid |17441|
|Private room_High   |1278 |
|Private room_Low    |8104 |
|Private room_Mid    |7714 |
|Shared room_High    |59   |
|Shared room_Low     |649  |
|Shared room_Mid     |204  |
+--------------------+-----+



### 3.6 Preparación de variables para Machine Learning

Los algoritmos de MLlib no aceptan variables categóricas directamente: requieren una representación numérica vectorial. Por este motivo se aplicaron las siguientes transformaciones:

1. **StringIndexer** sobre `Room Type`, `Property Type` y la variable objetivo `price_range`, convirtiendo cada categoría en un índice entero.
2. **OneHotEncoder** sobre los índices de las predictoras categóricas, para que el modelo no interprete el orden como una relación numérica.
3. **VectorAssembler** para ensamblar las variables predictoras (numéricas y categóricas) en una columna `features` de tipo vectorial.
4. **StandardScaler**  que estandariza las variables a media cero y varianza unitaria, para K-means porque esta basado en distancia euclidiana como K-Means.

## 4. Preparación de los conjuntos de entrenamiento y prueba

El 80% de la muestra M se utiliza en el entrenamiento y el 20% en el conjunto de prueba. Utilizamos la regla 80/20 (principio de Pareto) debido a que en muestras del orden  de decenas de miles, proporciona un balance aceptado en la practica ya que asgina suficientes observadores al entrenamiento para que los estimadores converjan en solucionbes estables y mantiene un conjunto prueba lo suficiente grande para que las métricas de desempeño tengan baja varianza muestral. La partición es pseudo-aleatoria con una semilla fija (`seed=42`) para asegurar la reproducibilidad de los experimentos. Separamos los datos en un conjunto de entrenamiento y uno de prueba para reducir el riesgo de **sobreajuste**, asegurando una estimación más confiable de su capacidad de generalización.

PySpark proporciona el método `randomSplit`, que realiza una partición aleatoria respetando las proporciones indicadas.

In [7]:
# División 80/20 train/test
train_df, test_df = muestra_M.randomSplit([0.8, 0.2], seed=42)

print("Tamaño entrenamiento:", train_df.count())
print("Tamaño prueba:       ", test_df.count())

# Distribución por price_range

print("\nDistribución price_range — Train:")
train_df.groupBy("price_range") \
        .count() \
        .orderBy("price_range") \
        .show()

print("Distribución price_range — Test:")
test_df.groupBy("price_range") \
       .count() \
       .orderBy("price_range") \
       .show()

# Distribución de los 9 estratos

print("\nEstratos Train:")
train_df.groupBy("Room Type", "price_range") \
        .count() \
        .orderBy("Room Type", "price_range") \
        .show(truncate=False)

print("Estratos Test:")
test_df.groupBy("Room Type", "price_range") \
       .count() \
       .orderBy("Room Type", "price_range") \
       .show(truncate=False)

Tamaño entrenamiento: 38665
Tamaño prueba:        9704

Distribución price_range — Train:
+-----------+-----+
|price_range|count|
+-----------+-----+
|       High| 9247|
|        Low| 9141|
|        Mid|20277|
+-----------+-----+

Distribución price_range — Test:
+-----------+-----+
|price_range|count|
+-----------+-----+
|       High| 2259|
|        Low| 2363|
|        Mid| 5082|
+-----------+-----+


Estratos Train:
+---------------+-----------+-----+
|Room Type      |price_range|count|
+---------------+-----------+-----+
|Entire home/apt|High       |8170 |
|Entire home/apt|Low        |2172 |
|Entire home/apt|Mid        |13970|
|Private room   |High       |1033 |
|Private room   |Low        |6446 |
|Private room   |Mid        |6140 |
|Shared room    |High       |44   |
|Shared room    |Low        |523  |
|Shared room    |Mid        |167  |
+---------------+-----------+-----+

Estratos Test:
+---------------+-----------+-----+
|Room Type      |price_range|count|
+---------------+-----

## 5. Construcción del modelo de aprendizaje supervisado (Random Forest)

Se eligió **Random Forest Classifier** por las siguientes características:

- **Robustez frente a outliers y escalas heterogéneas**: no requiere normalización previa dbeido a que esta basado en particiones del espacio.
- **Buen manejo de relaciones no lineales**: entre variables predictoras y la variable objetivo.
- **Reducción de varianza por *bagging***: se controla el sobreajuste de los arboles individuales porque promedia múltiples árboles entrenados sobre muestras (bootstrap).
- **Escalabilidad en PySpark**: MLlib implementa Random Forest de forma distribuida.

### 5.1 Variable objetivo y predictoras

- **Variable objetivo:** `price_range` (categórica: Low, Mid, High).
- **Predictoras numéricas:** `Bedrooms`, `Bathrooms`, `Accommodates`, `Beds`, `Cleaning Fee`, `Extra People`, `Guests Included`, `log_reviews`, `Host Listings Count`.
- **Predictoras categóricas:** `Room Type`, `Property Type`, `Cancellation Policy`, `Bed Type`.

### 5.2 Proceso de indexación, codificación, ensamblado y evaluación

Se creo un pipeline en el siguiente orden:
1. StringIndexer (para variables categóricas predictoras y variable objetivo price_range),
2. OneHotEncoder,
3. VectorAssembler (genera la columna rf_features),
4. RandomForestClassifier (200 árboles con profundidad máxima 15).

Para la evaluación se utilizaron dos métricas : **Accuracy** para la interpretabilidad y **F1-Score** para la robustez frente al desbalance entre clases.

In [8]:
# Variables categóricas — añadimos Cancellation Policy y Bed Type
cat_cols = ["Room Type", "Property Type", "Cancellation Policy", "Bed Type"]

# Variables numéricas — Cleaning Fee es la feature más correlacionada (r=0.55)
# Se eliminan Review Scores Rating y Availability 365 (r≈0.03, ruido puro)
num_cols_rf = [
    "Bedrooms", "Bathrooms", "Accommodates", "Beds",
    "Cleaning Fee", "Extra People", "Guests Included",
    "log_reviews", "Host Listings Count"
]

# Stages de preprocesamiento (compartidos por RF y GBT)
indexers = [
    StringIndexer(inputCol=c, outputCol=c.replace(" ","_")+"_idx", handleInvalid="keep")
    for c in cat_cols
]
encoders = [
    OneHotEncoder(inputCol=c.replace(" ","_")+"_idx", outputCol=c.replace(" ","_")+"_ohe")
    for c in cat_cols
]
label_indexer = StringIndexer(inputCol="price_range", outputCol="label", handleInvalid="keep")
assembler = VectorAssembler(
    inputCols=[c.replace(" ","_")+"_ohe" for c in cat_cols] + num_cols_rf,
    outputCol="rf_features",
    handleInvalid="skip"
)

preproc_stages = indexers + encoders + [label_indexer, assembler]

# Pipeline Random Forest
rf = RandomForestClassifier(
    labelCol="label", featuresCol="rf_features",
    numTrees=200, maxDepth=15,
    minInstancesPerNode=5,
    featureSubsetStrategy="sqrt",
    seed=42
)
pipeline_rf = Pipeline(stages=preproc_stages + [rf])


In [9]:
# Entrenar Random Forest
model_rf = pipeline_rf.fit(train_df)
pred_rf   = model_rf.transform(test_df)

# Recuperar etiquetas del label para la matriz de confusión
label_im = next(
    s for s in model_rf.stages
    if isinstance(s, StringIndexerModel) and s.getInputCol() == "price_range"
)
label_names = label_im.labelsArray[0]          # ej: ['Mid', 'Low', 'High']
idx_to_label = {float(i): lb for i, lb in enumerate(label_names)}

print("=" * 52)
print("  RANDOM FOREST")
print("=" * 52)
for metric in ["accuracy", "f1"]:
    score = MulticlassClassificationEvaluator(
        labelCol="label", predictionCol="prediction",
        metricName=metric).evaluate(pred_rf)
    print(f"  {metric.upper():10s} = {score:.4f}")

# Matriz de confusión con etiquetas legibles
to_label_udf = F.udf(lambda x: idx_to_label.get(x, str(x)))
pred_rf_lbl = (
    pred_rf
    .withColumn("true_label", to_label_udf(col("label")))
    .withColumn("pred_label", to_label_udf(col("prediction")))
)
print("\nMatriz de confusión (RF):")
pred_rf_lbl.groupBy("true_label").pivot("pred_label").count().orderBy("true_label").show()

# Importancia de variables
rf_model     = model_rf.stages[-1]
ohe_names    = []
for c in cat_cols:
    stage = next(
        s for s in model_rf.stages
        if isinstance(s, StringIndexerModel) and s.getInputCol() == c
    )
    lbls = stage.labelsArray[0]
    ohe_names += [f"{c}={lb}" for lb in lbls[:-1]]

feature_names = ohe_names + num_cols_rf
print("\nImportancia de variables (RF):")
for nm, v in sorted(
    zip(feature_names, rf_model.featureImportances.toArray()),
    key=lambda x: -x[1]
):
    print(f"  {nm:<42} {v:.4f}")

# Pipeline GBT + OneVsRest
print("\n" + "=" * 52)
print("  GRADIENT BOOSTED TREES (OneVsRest)")
print("=" * 52)

gbt = GBTClassifier(
    labelCol="label", featuresCol="rf_features",
    maxIter=100, maxDepth=8,
    stepSize=0.1, subsamplingRate=0.8,
    seed=42
)
ovr = OneVsRest(classifier=gbt, labelCol="label", featuresCol="rf_features")
pipeline_gbt = Pipeline(stages=preproc_stages + [ovr])

model_gbt = pipeline_gbt.fit(train_df)
pred_gbt  = model_gbt.transform(test_df)

for metric in ["accuracy", "f1"]:
    score = MulticlassClassificationEvaluator(
        labelCol="label", predictionCol="prediction",
        metricName=metric).evaluate(pred_gbt)
    print(f"  {metric.upper():10s} = {score:.4f}")

pred_gbt_lbl = (
    pred_gbt
    .withColumn("true_label", to_label_udf(col("label")))
    .withColumn("pred_label", to_label_udf(col("prediction")))
)
print("\nMatriz de confusión (GBT):")
pred_gbt_lbl.groupBy("true_label").pivot("pred_label").count().orderBy("true_label").show()


  RANDOM FOREST
  ACCURACY   = 0.6814
  F1         = 0.6765

Matriz de confusión (RF):
+----------+----+----+----+
|true_label|High| Low| Mid|
+----------+----+----+----+
|      High|1229|  72| 958|
|       Low|   7|1337|1019|
|       Mid| 354| 682|4046|
+----------+----+----+----+


Importancia de variables (RF):
  Host Listings Count                        0.2637
  Room Type=Entire home/apt                  0.1222
  Room Type=Private room                     0.0930
  Cleaning Fee                               0.0862
  Guests Included                            0.0855
  Extra People                               0.0604
  log_reviews                                0.0446
  Property Type=House                        0.0157
  Property Type=Apartment                    0.0145
  Property Type=Bed & Breakfast              0.0109
  Cancellation Policy=moderate               0.0080
  Cancellation Policy=moderate_new           0.0075
  Bed Type=Couch                             0.0049
  Cancel

### 5.5 Interpretación de resultados

Con **Random Forest** se obtuvo una **Accuracy de 0.6814** y un **F1-Score de 0.6765**. El modelo **GBT (OneVsRest)** superando ligeramente a RF con una **Accuracy de 0.6895** y un **F1-Score de 0.6863**. Ambos valores son sustancialmente superiores al 0.52 que alcanzaría un clasificador trivial que siempre predijera la **clase mayoritaria (Mid)**, por lo que el modelo captura información genuina de las variables predictoras. Ambos modelos presentan dificultad para distinguir entre las clases **High** y **Mid**, lo que es esperable dado que el límite entre ambas categorías ($p_{75}$ = 150 USD) es continuo.

El vector que muestra la importancia de variables no muestra que variables son más determinantes. Las más determinantes fueron:
- Host Listings Count                        0.2637
-   Room Type=Entire home/apt                  0.1222
-   Room Type=Private room                     0.0930
-   Cleaning Fee                               0.0862
-  Guests Included                            0.0855
-  Extra People                               0.0604
  
A diferencia de lo que se esperaba, variables de capacidad como Accommodates, Bedrooms y Bathrooms tuvieron importancias muy bajas (0.0005, 0.0034 y 0.0006 respectivamente), lo cual indica que el tipo de publicación y el comportamiento del anfitrión son más determinantes del rango de precio que el tamaño físico de la propiedad.

## 6. Modelo de aprendizaje no supervisado (K-Means)

**K-Means** es sencillo y escalable en entornos distribuidos. PySpark MLlib implementa la variante `k-means||`, permite descubrir perfiles latentes de propiedades sin etiquetas, complementando el enfoque supervisado.

### 6.1 Variables utilizadas

Variables numéricas: `Bedrooms`, `Bathrooms`, `Accommodates`, `Beds`, `Cleaning Fee`, `Extra People`, `Guests Included`, `log_reviews`, `Host Listings Count` y `Price`.

Debido a que K-Means opera con distancias euclidianas, se aplica `StandardScaler` para estandarizar y así ninguna variable domine por escala.

In [10]:
# Pipeline no supervisado: K-Means
# Features ampliadas con las variables de mayor correlación con el precio
cluster_features = [
    "Bedrooms", "Bathrooms", "Accommodates", "Beds",
    "Cleaning Fee", "Extra People", "Guests Included",
    "log_reviews", "Host Listings Count", "Price"
]

assembler_km = VectorAssembler(
    inputCols=cluster_features,
    outputCol="features_raw",
    handleInvalid="skip"
)
scaler = StandardScaler(
    inputCol="features_raw", outputCol="features",
    withMean=True, withStd=True
)

# Método del codo: evaluar k = 2 hasta 7
print("Método del codo:")
print(f"{'k':>3}  {'WSSSE':>14}  {'Silhouette':>12}")
print("-" * 36)

best_k, best_sil = 3, -1.0
for k in range(2, 8):
    km_tmp = KMeans(k=k, seed=42, featuresCol="features",
                    predictionCol="cluster", maxIter=20)
    pipe_tmp = Pipeline(stages=[assembler_km, scaler, km_tmp])
    m_tmp    = pipe_tmp.fit(muestra_M)
    p_tmp    = m_tmp.transform(muestra_M)
    wssse    = m_tmp.stages[-1].summary.trainingCost
    sil      = ClusteringEvaluator(
        featuresCol="features", predictionCol="cluster",
        metricName="silhouette", distanceMeasure="squaredEuclidean"
    ).evaluate(p_tmp)
    marker = "  ◄ mejor" if sil > best_sil else ""
    if sil > best_sil:
        best_sil = sil
        best_k   = k
    print(f"{k:>3}  {wssse:>14.2f}  {sil:>12.4f}{marker}")

print(f"\nk óptimo (mayor Silhouette): k = {best_k}")

# Modelo final con k óptimo
kmeans_final = KMeans(
    k=best_k, seed=42, featuresCol="features",
    predictionCol="cluster", maxIter=50
)
pipeline_km = Pipeline(stages=[assembler_km, scaler, kmeans_final])
model_km    = pipeline_km.fit(muestra_M)
pred_km     = model_km.transform(muestra_M)

silhouette = ClusteringEvaluator(
    featuresCol="features", predictionCol="cluster",
    metricName="silhouette", distanceMeasure="squaredEuclidean"
).evaluate(pred_km)
print(f"Silhouette Score (k={best_k}) = {silhouette:.4f}")

Método del codo:
  k           WSSSE    Silhouette
------------------------------------
  2       364216.88        0.6472  ◄ mejor
  3       331577.29        0.4846
  4       297711.29        0.2716
  5       269631.53        0.6460
  6       240143.90        0.4802
  7       237865.68        0.4770

k óptimo (mayor Silhouette): k = 2
Silhouette Score (k=2) = 0.6472


Se evaluaron valores de k de 2 a 7. El criterio de selección fue el mayor **Silhouette Score**, donde en el rango $[-1, 1]$: valores cercanos a 1 indican clusters bien separados. El valor de **k = 2** resultó óptimo al maximizar el Silhouette Score (0.6472).

In [11]:
# Perfil descriptivo de cada cluster
pred_km.groupBy("cluster").agg(
    F.count("*").alias("n"),
    F.round(F.mean("Price"),            2).alias("Price_mean"),
    F.round(F.mean("Bedrooms"),         2).alias("Bedrooms_mean"),
    F.round(F.mean("Bathrooms"),        2).alias("Bathrooms_mean"),
    F.round(F.mean("Accommodates"),     2).alias("Accommodates_mean"),
    F.round(F.mean("Beds"),             2).alias("Beds_mean"),
    F.round(F.mean("Cleaning Fee"),     2).alias("CleaningFee_mean"),
    F.round(F.mean("Number of Reviews"),2).alias("Reviews_mean"),
    F.round(F.mean("Host Listings Count"),2).alias("HostListings_mean"),
    F.round(F.mean("Availability 365"), 2).alias("Avail365_mean")
).orderBy("cluster").show()

# Cruce con price_range para validar coherencia con el modelo supervisado
print("Cruce cluster vs. price_range:")
pred_km.groupBy("cluster").pivot("price_range").count().orderBy("cluster").show()

+-------+-----+----------+-------------+--------------+-----------------+---------+----------------+------------+-----------------+-------------+
|cluster|    n|Price_mean|Bedrooms_mean|Bathrooms_mean|Accommodates_mean|Beds_mean|CleaningFee_mean|Reviews_mean|HostListings_mean|Avail365_mean|
+-------+-----+----------+-------------+--------------+-----------------+---------+----------------+------------+-----------------+-------------+
|      0|39636|    104.03|         1.06|          1.09|              2.6|     1.45|            40.1|       16.99|             5.39|       161.51|
|      1| 8733|    295.71|         2.71|           1.9|             6.34|     3.97|          101.13|        15.2|             28.9|       183.26|
+-------+-----+----------+-------------+--------------+-----------------+---------+----------------+------------+-----------------+-------------+

Cruce cluster vs. price_range:
+-------+----+-----+-----+
|cluster|High|  Low|  Mid|
+-------+----+-----+-----+
|      0|55

### 6.2 Interpretación de los grupos obtenidos

Como $k=2$ resultó óptimo al maximizar el Silhouette Score, eso nos indica que el modelo identificó dos clusters con perfiles claramente diferenciados:

- **Cluster 0: Económico-Estándar** (n = 39,636): propiedades pequeñas con precio promedio de **104 USD**, aproximadamente 1 habitación, entre 1 y 2 camas, 1 baño y capacidad para 2-3 personas, con disponibilidad media de 161 días al año. Concentrando la mayoría de propiedades `Low` (11,307) y `Mid` (22,759).
- **Cluster 1: Premium** (n = 8,733): propiedades más grandes con precio promedio de **296 USD**, aproximadamente 3 habitaciones, 2 baños y capacidad para 6 o más personas, con tarifa de limpieza media de $101 USD y alta disponibilidad (183 días). Concentra la mayoría de propiedades `High` (5,936).
---

## 7. Discusión de resultados

El desempeño de los modelos supervisados es consistente entre ambos enfoques, con una ligera ventaja del **Gradient Boosted Trees (GBT)** frente al **Random Forest**, lo cual indica que la relación entre variables y rango de precio presenta cierto grado de no linealidad.

La matriz de confusión evidencia que los principales errores se concentran en la frontera entre las clases **Mid y High**, lo cual es coherente con la naturaleza continua del precio y la discretización utilizada para construir la variable objetivo. Mientras que la clase **Low** está más separada, lo que indica que las propiedades económicas están mejor definidas en el espacio de características.

Por otro lado, el valor óptimo de **k = 2**, indica la presencia de dos estructuras principales en el mercado de alojamientos. Identificando 2 clústers:

- Propiedades económicas y estándar con alta densidad.
- Propiedades premium con mayor capacidad, mayor precio y mayor tamaño.

La combinación de modelos supervisados y no supervisados nos da una visión más completa del mercado inmobiliario. De tal forma que un inversionista puede identificar el cluster al que pertenece una propiedad, estimar su rango de precio esperado y comparar su posición frente a propiedades similares. Además, la segmentación permite detectar zonas o tipos de propiedades con mayor potencial de rentabilidad.

### 7.1 Áreas de oportunidad

- El uso de una muestra del 10% reduce el costo computacional, pero introduce variabilidad muestral.
- La discretización del precio en terciles implica pérdida de información y puede influir en la frontera entre clases.
- No se incorporan variables geoespaciales detalladas, que pueden ser altamente predictivas.

---
## 8. Conclusiones

El presente trabajo demuestra la eficacia de ySpark MLlib para resolver problemas de aprendizaje automático en entornos de Big Data, utilizando como caso de estudio de Airbnb.

Los principales hallazgos son:

1. El rango de precio de una propiedad puede predecirse con buen desempeño a partir de sus características estructurales. Los modelos Random Forest y GBT alcanzan métricas superiores a un clasificador trivial.
2. Las variables más influyentes corresponden a aspectos relacionados con el tipo de propiedad y el comportamiento del anfitrión.
3. El mercado de alojamientos presenta una estructura segmentada en dos grupos principales, correspondientes a perfiles económicos-estándar y premium. E
4. PySpark MLlib permite construir pipelines completos de machine learning escalables, capaces de procesar grandes volúmenes de datos de manera eficiente.
5. La combinación de modelos supervisados y no supervisados nos da una visión más completa del caso de estudio.


# Referencias

[1] Polak, A. (2023). *Scaling Machine Learning with Spark: Distributed ML with MLlib, TensorFlow, and PyTorch*. Chapter 4: *Data Ingestion, Preprocessing and Descriptive Statistics*. O'Reilly Media.

[2] Polak, A. (2023). *Scaling Machine Learning with Spark: Distributed ML with MLlib, TensorFlow, and PyTorch*. Chapter 6: *Supervised Machine Learning* y *Unsupervised Machine Learning*. O'Reilly Media.

[3] Apache Spark. *PySpark ML API Documentation*. Disponible en: https://spark.apache.org/docs/latest/api/python/reference/pyspark.ml.html

[4] Apache Spark. *MLlib Statistics Documentation*. Disponible en: https://spark.apache.org/docs/latest/ml-statistics.html

[5] Apache Spark. *MLlib Classification and Regression Guide*. Disponible en: https://spark.apache.org/docs/latest/ml-classification-regression.html

[6] Apache Spark. *MLlib Clustering Guide*. Disponible en: https://spark.apache.org/docs/latest/ml-clustering.html

[7] JoeBeachCapital. (s.f.). Airbnb dataset. Kaggle. https://www.kaggle.com/datasets/joebeachcapital/airbnb